sklearn/impute.py

```diff
@@ -629,3 +632,262 @@ def fit_transform(self, X, y=None):
+                             "with integer dtype or an array of string values "
+                             "with an object dtype.".format(X.dtype))
+
+        return X
+
+    def _vect_is_not_none(self, a):
+        def is_not_none(x):
+            return x is not None
+
+        return np.vectorize(is_not_none)(a)
+
+    def _get_values_counts(self, array):
+        if array.size > 0:
+            counter = Counter(array)
+            values, counts = zip(*counter.items())
+            values = np.array(values)
+            counts = np.array(counts, dtype=float)
+        else:
+            values = np.empty(0)
+            counts = np.empty(0)
+        return values, counts
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : SamplingImputer
+        """
+        self._random_state_ = getattr(self, "random_state_",
+                                      check_random_state(self.random_state))
+
+        X = self._validate_input(X)
+
+        if sparse.issparse(X):
+            # missing_values = 0 not allowed with sparse data as it would
+            # force densification
+            if self.missing_values == 0:
+                raise ValueError("Imputation not possible when missing_values"
+                                 " == 0 and input is sparse. Provide a dense "
+                                 "array instead.")
+            else:
+                self.uniques_, self.probas_ = \
+                        self._sparse_fit(X, self.missing_values)
+        else:
+            self.uniques_, self.probas_ = self._dense_fit(X,
+                                                          self.missing_values)
+
+        return self
+
+    def _sparse_fit(self, X, missing_values):
+        """Fit the transformer on sparse data."""
+        mask_data = _get_mask(X.data, missing_values)
+        n_implicit_zeros = X.shape[0] - np.diff(X.indptr)
+
+        uniques = np.empty(X.shape[1], dtype=object)
+        probas = np.empty(X.shape[1], dtype=object)
+
+        for i in range(X.shape[1]):
+            column = X.data[X.indptr[i]:X.indptr[i+1]]
+            mask_column = mask_data[X.indptr[i]:X.indptr[i+1]]
+            column = column[~mask_column]
+
+            values, counts = self._get_values_counts(column)
+
+            # count implicit zeros
+            if n_implicit_zeros[i] > 0:
+                if 0 in values:
+                    counts[values == 0] += n_implicit_zeros[i]
+                else:
+                    values = np.append(values, 0)
+                    counts = np.append(counts, n_implicit_zeros[i])
+
+            if values.size > 0:
+                uniques[i] = values
+                if values.size == X.shape[0]:
```

Comment:
This should never really happen in the sparse case. I don't think we need to support it.

Reply:
Well this brings no overhead so why not keep it ?